web-api/src/mind.rs

```diff
@@ -62,20 +66,77 @@ fn run_benchmark() -> Result<(), anyhow::Error> {
         .map(|result| result.map(|article: Article| (article.id.clone(), article)))
         .collect::<Result<HashMap<_, _>, _>>()?;
 
-    let impressions_iter = read("behaviors.tsv")?;
+    let impressions = read("behaviors.tsv")?;
+
+    let nranks = vec![3];
+    let mut ndcgs = Array::zeros((nranks.len(), 0));
 
-    for impression in impressions_iter {
+    // Loop over all impressions, prepare reranker with news in click history
+    // and rerank the news in an impression
+    for impression in impressions {
         let impression: Impression = impression?;
-        let clicks = impression.clicks.split(' ').collect::<Vec<&str>>();
-        match articles.get(clicks[0]) {
-            Some(imp) => println!("{:?}", imp),
-            None => println!("Article id {} not found.", clicks[0]),
-        }
+        let clicks_iter = impression.clicks.split(' ');
+
+        // Placeholder for interacting with the entire click history
+        clicks_iter.for_each(|click| match articles.get(click) {
+            Some(article) => println!("The article {:?} was interacted.", article),
+            None => println!("Article id {} not found.", click),
+        });
+
+        let news = impression.news.split(' ').collect::<Vec<&str>>();
+
+        // Placeholder for reranking the results
+        let news_labels_iter = news.iter().map(|x| x.split('-').collect::<Vec<_>>());
+
+        let mut snippet_label_pairs = news_labels_iter
+            .map(|id_label| match articles.get(id_label[0]) {
+                Some(article) => {
+                    SnippetLabelPair((*article.snippet).to_string(), id_label[1].to_string())
+                }
+                _ => unreachable!(),
+            })
+            .collect::<Vec<_>>();
+        snippet_label_pairs.shuffle(&mut thread_rng());
+
+        let labels = snippet_label_pairs
+            .iter()
+            .map(|snippet_label| snippet_label.1.parse::<f32>().unwrap())
+            .collect::<Vec<_>>();
+        let ndcgs_iteration = ndcg(&labels[..], &nranks);
+
+        ndcgs
+            .push_column(ArrayView::from(&ndcgs_iteration))
+            .unwrap();
     }
+    println!("{:?}", ndcgs);
 
     Ok(())
 }
 
+fn ndcg(relevance: &[f32], k: &[usize]) -> Vec<f32> {
+    let mut optimal_order = relevance.to_owned();
+    optimal_order.sort_by(|a, b| b.partial_cmp(a).unwrap());
```

Comment:
:art: switching the arguments to sort them descendingly does the work, but it can easily be overlooked. an alternative is to use the `reverse()` method on the `Ordering` to make it more visible.
```suggestion
    optimal_order.sort_by(|a, b| a.partial_cmp(b).unwrap().reverse());
```

:wrench: in this case to avoid panics we have some utility to sort floats in the `xayn_ai_coi` crate (i admit it's not placed well there), namely `nan_safe_f32_cmp()` and `nan_safe_f32_cmp_desc()` (which can be used here without repeating the arguments):
```suggestion
    optimal_order.sort_by(nan_safe_f32_cmp_desc);
```

Reply:
Agree with the overlooking. Both suggestions make sense. Good to know u have it already. Will stick to the second one.